greenplumpython/expr.py

```diff
@@ -7,9 +7,15 @@
 
 
 class Expr:
-    def __init__(self, as_name: Optional[str] = None, db: Optional[Database] = None) -> None:
+    def __init__(
+        self,
+        as_name: Optional[str] = None,
+        table: Optional["Table"] = None,
+        db: Optional[Database] = None,
```

Comment:
I have a question:

In fact, we did not support multiple database in one SQL. and dblink need some special API to do cross database query.

why not put database in `fetch()`.

or some context API (or session API), the may API looks:

```py
gp.GlobalContext = gp.ContextBuilder().WithDatabase("xxxx");

"some very long SQL build".fetch();

or

"some very long SQL build".fetch(gp.GlobalContext);
```

Reply:
Great  question! Even though we cannot access multiple database in one SQL statement, I think it is perfectly valid if the user wants to use GreenplumPython as a "pipe" to transmit data between any two databases. For example, 
```python
t1 = gp.table("t1", db=db1)
data = t1.fetch()
t2 = gp.table("t2", db=db2)
t2.insert(data)
```

Reply:
> Great question! Even though we cannot access multiple database in one SQL statement, I think it is perfectly valid if the user wants to use GreenplumPython as a "pipe" to transmit data between any two databases. For example,

in session API:

```python
data = gp.table("t1").fetch(gp.ContextBuilder().WithDatabase("db1"))
gp.table("t2", db=db2).insert().execute(gp.ContextBuilder().WithDatabase("db2"))
```

Reply:
Why do you want to introduce this `ContextBuilder()`?

Reply:
> Why do you want to introduce this `ContextBuilder()`?

not introduce `ContextBuilder`, think some global connection will be better, the user doesn't need to remember which table has a database and which table don't.

the API can be `ContextBuilder()` or magic string (`%sql`) in ipyhon, or `SparkSession` in pyspark 
```python
from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local[1]").appName("SparkByExamples.com").getOrCreate() 
```

Reply:
and the hierarchy, current:

```
Expr         = Table      + Database
BinaryExpr   = Expr       + Operator   + Expr
FunctionCall = "Function" + Database
Column       = "Column"   + Table
Table        = "Query"    + [Table]    + "Name" + Database
```

the hierarchy does not very clear, just simple and works.
I think we should follow the SQL AST to order our hierarchy.

for example, in `SELECT` `DISTINCT` is a keyword, not an expression (or an operator)

```
[ WITH [ RECURSIVE ] with_query [, ...] ]
SELECT [ ALL | DISTINCT [ ON ( expression [, ...] ) ] ]
    [ * | expression [ [ AS ] output_name ] [, ...] ]
    [ FROM from_item [, ...] ]
    [ WHERE condition ]
    [ GROUP BY [ ALL | DISTINCT ] grouping_element [, ...] ]
    [ HAVING condition ]
    [ WINDOW window_name AS ( window_definition ) [, ...] ]
    [ { UNION | INTERSECT | EXCEPT } [ ALL | DISTINCT ] select ]
    [ ORDER BY expression [ ASC | DESC | USING operator ] [ NULLS { FIRST | LAST } ] [, ...] ]
    [ LIMIT { count | ALL } ]
    [ OFFSET start [ ROW | ROWS ] ]
    [ FETCH { FIRST | NEXT } [ count ] { ROW | ROWS } { ONLY | WITH TIES } ]
    [ FOR { UPDATE | NO KEY UPDATE | SHARE | KEY SHARE } [ OF table_name [, ...] ] [ NOWAIT | SKIP LOCKED ] [...] ]
```
it means we can not add multiple `DISTINCT` in one SELECT part, `DISTINCT` is a part of `SelectStmt`.

so the API for `SelectStmt` looks like:

```
.all(exp: [Expression])                   -> SelectStmt(without all and distinct )
.distinct(exp: [Expression])              -> SelectStmt(without all and distinct )
.from([exp: Expression, as: Str])         -> SelectStmt(without from)
.where(con: Condition)                    -> SelectStmt(without where)
.GroupBy(grouping_element: [str])         -> SelectStmt(without groupby groupbyall groupbydistinct)
.GroupByAll(grouping_element: [str])      -> SelectStmt // pg9 not support this
.GroupByDistinct(grouping_element: [str]) -> SelectStmt // pg9 not support this
.Having(con: condition)                   -> SelectStmt(whiout having)

and more things like that format

```

the final thing our user needs to do does not change with this wrap.

```
gp.table(tablename: str)     -> from_item // a subtype of table, same as new_table in InsertStmt
from_item.fetch()            -> Iter      // wrap with select * from <table>
expression.fetch()           -> Iter      // wrap with select * from <expression>
from_item.insert(expression) -> void      // insert into <from_item> values <expression>

user can still do:

t1 = gp.table("t1", db=db1)
data = t1.fetch()
t2 = gp.table("t2", db=db2)
t2.insert(data)
```

here is PostgreSQL's expression type https://www.postgresql.org/docs/current/sql-expressions.html#SQL-EXPRESSIONS-COLUMN-REFS
and here is AST for SELECT https://www.postgresql.org/docs/9.6/sql-select.html
also AST for INSERT https://www.postgresql.org/docs/9.6/sql-insert.html

Reply:
Thanks for your feedback! Will consider these issues.